### PR TITLE
Keychron K14 bring up

### PR DESCRIPTION
**Pull request by Jpe230**
Add initial support for the newly released Keychron K14 RGB

## Description

Added a default keymap that mimics the original firmware, all features seem to work (except bluetooth) and the keyboard is very stable with QMK

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

### CODE REVIEW

File: keyboards/keychron/k14/halconf.h [hunk 1]
```diff
@@ -0,0 +1,524 @@
...
+ */
+#    if !defined(HAL_USE_MMC_SPI) || defined(__DOXYGEN__)
+#        define HAL_USE_MMC_SPI FALSE
+#    endif
+
+/**
+ * @brief   Enables the PWM subsystem.
+ */
+#    if !defined(HAL_USE_PWM) || defined(__DOXYGEN__)
+#        define HAL_USE_PWM FALSE
```
Comment (dexter93) [inline keyboards/keychron/k14/halconf.h]: set this to true and remove the line from mcuconf.h

File: keyboards/keychron/k14/config.h [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "config_common.h"
+
+/* USB Device descriptor parameter */
+#define VENDOR_ID       0x05AC
+#define PRODUCT_ID      0x024F
```
Comment (dexter93) [inline keyboards/keychron/k14/config.h]: This is not accepted anymore for new keyboards. Please see https://github.com/SonixQMK/qmk_firmware/issues/152

File: keyboards/keychron/k14/config_led.h [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+#define LED_MATRIX_ROWS 5
+#define LED_MATRIX_ROW_CHANNELS 3
+#define LED_MATRIX_ROWS_HW (LED_MATRIX_ROWS * LED_MATRIX_ROW_CHANNELS)
+//#define LED_MATRIX_ROW_PINS   { -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1 }
+#define LED_MATRIX_ROW_PINS { C6, C5, C4, C9, C8, C7, C12, C11, C10, B13, C14, C13, B14, B15, D3 }
+
+#define LED_MATRIX_COLS 16
+//#define LED_MATRIX_COL_PINS { -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1}
+#define LED_MATRIX_COL_PINS { A8, A9, A10, A11, A12, A13, A14, A15, B0, B1, B2, B3, B4, B5, B6, B7}
```
Comment (dexter93) [inline keyboards/keychron/k14/config_led.h]: you can remove the double definition by simply doing

`#define LED_MATRIX_COL_PINS MATRIX_COL_PINS`

File: keyboards/keychron/k14/bootloader_defs.h [hunk 4]
```diff
@@ -0,0 +1,3 @@
+/* Address for jumping to bootloader on SN32 chips. */
+/* It is chip dependent */
+// #define SN32_BOOTLOADER_ADDRESS 0x1FFF0000
```
Comment (dexter93) [inline keyboards/keychron/k14/bootloader_defs.h]: This file is no longer required

File: keyboards/keychron/k14/config.h [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+#define MATRIX_ROW_PINS { D11, D10, D9, D8, D7 }
+
+// Connects each switch in the dip switch to the GPIO pin of the MCU
+#define DIP_SWITCH_PINS { D4, D5 }
+
+/* Debounce reduces chatter (unintended double-presses) - set 0 if debouncing is not needed */
+#define DEBOUNCE 0
+
+/* LED Status indicators */
+//#define LED_CAPS_LOCK_PIN B9
```
Comment (dexter93) [inline keyboards/keychron/k14/config.h]: I would expect the K14 to come with a dedicated LED for Caps Lock. If so, please map it here

File: keyboards/keychron/k14/config.h [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+#define MATRIX_ROWS 5
+#define MATRIX_COLS 16
+
+#define DIODE_DIRECTION COL2ROW
+
+#define MATRIX_COL_PINS { A8, A9, A10, A11, A12, A13, A14, A15, B0, B1, B2, B3, B4, B5, B6, B7 }
+#define MATRIX_ROW_PINS { D11, D10, D9, D8, D7 }
+
+// Connects each switch in the dip switch to the GPIO pin of the MCU
+#define DIP_SWITCH_PINS { D4, D5 }
```
Comment (dexter93) [inline keyboards/keychron/k14/config.h]: D4 should be the MAC switch signal and D5 should be the BT switch signal. These look like a copy paste job. Please make sure they are correct

**Review by dexter93 — CHANGES_REQUESTED** [review]
Hi, went through the PR. Nice job on the bringup. Please address the comments and cleanup any temporary commented code so we can merge this in

File: keyboards/keychron/k14/config.h (hunk 6 above)
Comment (Jpe230) [reply keyboards/keychron/k14/config.h]: Encountered a bug: if user changes to BT mode then the RGB matrix lags, resolving since the mapping is correct.

**Pull request merged**